File: examples/containers_example.cpp
/**
 * @file containers_example.cpp
 * @brief Example demonstrating the usage of the containers module
 */

#include <algorithm>
#include <cstddef>
#include <format>
#include <print>
#include <string>
#include <vector>

#include "containers/container.hpp"

using cpp_features::containers::Container;

auto main() -> int {
  std::println("=== Containers Module Example ===");

  // Demonstrate empty container
  Container<int> empty_container;
  std::println("Empty container: {}", empty_container);

  // Demonstrate initialization with integers
  Container<int> numbers{1, 2, 3, 4, 5};
  std::println("Integer container: {}", numbers);

  // Demonstrate adding elements
  numbers.Add(6);
  numbers.Add(7);
  std::println("After adding elements: {}", numbers);

  // Demonstrate removing elements
  auto value = 3;
  auto removed_count = numbers.Remove(value);
  std::println("Removed {} elements of value {}", removed_count, value);
  std::println("After removing elements: {}", numbers);

  // Demonstrate safe access to elements
  std::size_t index = 3;
  auto result = numbers.At(index);
  if (result) {
    std::println("Element at index {}: {}", index, result->get());
  } else {
    std::println("Error accessing element at index {}", index);
  }

  // Demonstrate views and filters
  auto even_view = numbers.GetFilteredView([](int x) { return x % 2 == 0; });
  std::print("Even numbers: ");
  std::ranges::for_each(even_view, [](int x) { std::print("{} ", x); });
  std::println();

  // Demonstrate transformation
  auto doubled_view = numbers.GetTransformedView([](int n) { return n * 2; });
  std::print("Doubled numbers: ");
  std::ranges::for_each(doubled_view, [](int x) { std::print("{} ", x); });
  std::println();

  // Demonstrate initialization with strings
  Container<std::string> words{"hello", "modern", "cpp", "world"};
  std::println("String container: {}", words);

  // Demonstrate emplacing elements
  words.Emplace(5, '!');
  std::println("After emplacing elements: {}", words);

  // Demonstrate initialization with doubles
  Container<double> prices{19.999, 29.95, 5.50};
  std::println("Floating point container: {}", prices);

  // Demonstrate CTAD from range
  std::vector source{10, 20, 30};
  Container container(source);
  std::println("CTAD from range: {}", container);

  std::println("=== Containers Module Example Completed ===");
  return 0;
}

File: python/examples/containers_example.py
"""Example demonstrating the usage of the containers module."""

import sys
from pathlib import Path

python_dir = Path(__file__).parent.parent
sys.path.append(str(python_dir / 'src'))
sys.path.append(str(python_dir.parent / 'build' / 'binding'))

from demo.containers import Container


def main() -> None:
    """Run all container examples."""
    print('=== Containers Module Example ===')

    # Demonstrate empty container
    empty_container = Container(int)
    print(f'Empty container: {empty_container}')

    # Demonstrate initialization with integers
    numbers = Container(int, [1, 2, 3, 4, 5])
    print(f'Integer container: {numbers}')

    # Demonstrate adding elements
    numbers.add(6)
    numbers.add(7)
    print(f'After adding elements: {numbers}')

    # Demonstrate removing elements
    value = 3
    removed_count = numbers.remove(value)
    print(f'Removed {removed_count} elements of value {value}')
    print(f'After removing elements: {numbers}')

    # Demonstrate safe access to elements
    index = 3
    try:
        element = numbers[index]
        print(f'Element at index {index}: {element}')
    except IndexError:
        print(f'Error accessing element at index {index}')

    # Demonstrate filtering
    even_numbers = numbers.filter(lambda x: x % 2 == 0)
    print(f'Even numbers: {even_numbers}')

    # Demonstrate transformation
    doubled_numbers = numbers.transform(lambda n: n * 2)
    print(f'Doubled numbers: {doubled_numbers}')

    # Demonstrate initialization with strings
    words = Container(str, ['hello', 'modern', 'python', 'world'])
    print(f'String container: {words}')

    # Demonstrate initialization with doubles
    prices = Container(float, [19.999, 29.95, 5.50])
    print(f'Floating point container: {prices}')

    print('=== Containers Module Example Completed ===')


if __name__ == '__main__':
    main()
